libraries/extractor/src/main/java/androidx/media3/extractor/ts/DtsUhdReader.java

```diff
@@ -0,0 +1,246 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package androidx.media3.extractor.ts;
+
+import static java.lang.Math.min;
+
+import androidx.annotation.Nullable;
+import androidx.media3.common.C;
+import androidx.media3.common.Format;
+import androidx.media3.common.ParserException;
+import androidx.media3.common.util.Assertions;
+import androidx.media3.common.util.ParsableByteArray;
+import androidx.media3.common.util.UnstableApi;
+import androidx.media3.common.util.Util;
+import androidx.media3.extractor.DtsUtil;
+import androidx.media3.extractor.ExtractorOutput;
+import androidx.media3.extractor.TrackOutput;
+import androidx.media3.extractor.ts.TsPayloadReader.TrackIdGenerator;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.RequiresNonNull;
+
+/** Parses a continuous DTS UHD byte stream and extracts individual samples. */
+@UnstableApi
+public final class DtsUhdReader implements ElementaryStreamReader {
+
+  private static final int STATE_FINDING_SYNC = 0;
+  private static final int STATE_FIND_HEADER_SIZE = 1;
+  private static final int STATE_READING_HEADER = 2;
+  private static final int STATE_READING_SAMPLE = 3;
+
+  private static final int FTOC_HEADER_SIZE_MIN = 7;
+  /**
+   * Maximum size of DTS UHD(DTS:X) frame header.
+   * See ETSI TS 103 491 V1.2.1 (2019-05) section 6.4.4.3.
+   * */
+  private static final int FTOC_HEADER_SIZE_MAX = 5408;
+
+  private final ParsableByteArray headerScratchBytes;
+  @Nullable private final String language;
+
+  private @MonotonicNonNull String formatId;
+  private @MonotonicNonNull TrackOutput output;
+
+  private int state;
+  private int bytesRead;
+
+  // Used to find the header.
+  private int syncBytes;
+
+  // Used when parsing the header.
+  private long sampleDurationUs;
+  private @MonotonicNonNull Format format;
+  private int sampleSize;
+  private boolean isFtocSync;
+  private int headerSizeToRead;
+
+  // Used when reading the samples.
+  private long timeUs;
+  private int sampleRate;
+  private int sampleCount; // frame duration
+
+  /**
+   * Constructs a new reader for DTS UHD elementary streams.
+   *
+   * @param language Track language.
+   */
+  public DtsUhdReader(@Nullable String language) {
+    headerScratchBytes = new ParsableByteArray(new byte[FTOC_HEADER_SIZE_MAX]);
+    state = STATE_FINDING_SYNC;
+    timeUs = C.TIME_UNSET;
+    sampleRate = 48000; // initialize to a non-zero sampling rate
```

Comment:
Why do we choose this value 48000 as the default?

Reply:
The sampleRate information will be updated only in the FTOC sync frame. For any subsequent FTOC non-sync frames, the sampleRate will be the same as the previous sync frame.
The sampleRate information is used in the below calculation.
 `sampleDurationUs = (int)(C.MICROS_PER_SECOND * sampleCount / sampleRate);`

By any chance, if the extractor encounters a non-sync frame at the very beginning, the sampleRate won't get extracted from the stream. To avoid 'division by zero' in the sampleDurationUs calculation, I have initialized the sampleRate value to a non-zero value. The UHD streams are encoded at 48000 sample rate. I used the same value for initializing sampleRate variable.